### PR TITLE
Fix BigFloat for Julia v1.12

### PR DESCRIPTION
**Pull request by odow**

### CODE REVIEW

File: src/implementations/BigFloat.jl [hunk 1]
```diff
@@ -9,12 +9,18 @@
 
 mutability(::Type{BigFloat}) = IsMutable()
 
-# Copied from `deepcopy_internal` implementation in Julia:
-# https://github.com/JuliaLang/julia/blob/7d41d1eb610cad490cbaece8887f9bbd2a775021/base/mpfr.jl#L1041-L1050
-function mutable_copy(x::BigFloat)
-    d = x._d
-    d′ = GC.@preserve d unsafe_string(pointer(d), sizeof(d)) # creates a definitely-new String
-    return Base.MPFR._BigFloat(x.prec, x.sign, x.exp, d′)
+@static if VERSION >= v"1.12"
```
Comment (odow) [inline src/implementations/BigFloat.jl:12]:
```suggestion
@static if VERSION >= v"1.12.0-DEV.1343"
```

File: src/implementations/BigFloat.jl [hunk 2]
```diff
@@ -9,12 +9,20 @@
...
-function mutable_copy(x::BigFloat)
-    d = x._d
-    d′ = GC.@preserve d unsafe_string(pointer(d), sizeof(d)) # creates a definitely-new String
-    return Base.MPFR._BigFloat(x.prec, x.sign, x.exp, d′)
+# These methods are copied from `deepcopy_internal` in `base/mpfr.jl`. We don't
+# use `mutable_copy(x) = deepcopy(x)` because this creates an empty `IdDict()`
+# which costs some extra allocations. We don't need the IdDict case because we
+# never call `mutable_copy` recursively.
+@static if VERSION >= v"1.12.0-DEV.1343"
+    mutable_copy(x::BigFloat) = Base.MPFR._BigFloat(copy(getfield(x, :d)))
```
Comment (odow) [inline src/implementations/BigFloat.jl:17]:
```suggestion
    # mutable_copy(x::BigFloat) = Base.MPFR._BigFloat(copy(getfield(x, :d)))
    mutable_copy(x::BigFloat) = deepcopy(x)
```

**Comment by odow** [comment]
I don't understand this well enough to figure out what's going on here.

**Comment by odow** [comment]
I guess a bunch of our C calls might be invalid now?
```julia
julia> x, y = setprecision(BigFloat, 32) do
           return rand(BigFloat, 1), rand(BigFloat, 1)
       end
(BigFloat[0.51086560986], BigFloat[0.18767778529])

julia> accurate = LinearAlgebra.dot(x, y)
0.095878126238614356830099183515159211310674436390399932861328125

julia> output = setprecision(BigFloat, 32) do
           buf = MA.buffer_for(LinearAlgebra.dot, Vector{BigFloat}, Vector{BigFloat})
           output = BigFloat(0)
           MA.buffered_operate_to!!(buf, output, LinearAlgebra.dot, x, y)
           return output
       end
get_str.c:157: MPFR assertion failed: size_s1 >= m

[48611] signal 6: Abort trap: 6
in expression starting at REPL[85]:1
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
pthread_kill at /usr/lib/system/libsystem_pthread.dylib (unknown line)
abort at /usr/lib/system/libsystem_c.dylib (unknown line)
mpfr_assert_fail at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
mpfr_get_str_aux at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
mpfr_get_str at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
regular_eg at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
sprnt_fp at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
mpfr_vasnprintf_aux at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
mpfr_asprintf at /Users/oscar/.julia/juliaup/julia-nightly/lib/julia/libmpfr.6.dylib (unknown line)
string_mpfr at ./mpfr.jl:1187
_string at ./mpfr.jl:1233
_string at ./mpfr.jl:1235 [inlined]
show at ./mpfr.jl:1245
show at ./multimedia.jl:47
unknown function (ip: 0x104e1d5e2)
```

This PR seems suspicous https://github.com/JuliaLang/julia/pull/55906

**Comment by blegat** [comment]
The printing is actually happened when we try to show the output. Surprisingly, replacing `output = BigFloat(0)` by `output = BigFloat(1)` removes the segfault.

**Comment by blegat** [comment]
Here is the MWE
```julia
a = one(BigFloat)
b = zero(BigFloat)
ccall((:mpfr_swap, :libmpfr), Cvoid, (Ref{BigFloat}, Ref{BigFloat}), b, a)
show(b)
```
It gives
```
get_str.c:157: MPFR assertion failed: size_s1 >= m

[42807] signal 6 (-6): Aborted
in expression starting at /home/blegat/.julia/dev/MutableArithmetics/b.jl:6
unknown function (ip: 0x725fea6533f4)
gsignal at /usr/lib/libc.so.6 (unknown line)
abort at /usr/lib/libc.so.6 (unknown line)
mpfr_assert_fail at /workspace/srcdir/mpfr-4.2.1/src/mpfr-gmp.c:300
mpfr_get_str_aux at /workspace/srcdir/mpfr-4.2.1/src/get_str.c:157
mpfr_get_str at /workspace/srcdir/mpfr-4.2.1/src/get_str.c:2877
regular_eg at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1308
partition_number at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1925 [inlined]
sprnt_fp at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:1994 [inlined]
mpfr_vasnprintf_aux at /workspace/srcdir/mpfr-4.2.1/src/vasprintf.c:2413
mpfr_asprintf at /workspace/srcdir/mpfr-4.2.1/src/printf.c:176
string_mpfr at ./mpfr.jl:1187
_string at ./mpfr.jl:1233
_string at ./mpfr.jl:1235 [inlined]
show at ./mpfr.jl:1245
unknown function (ip: 0x725fe8d03866)
show at ./show.jl:476
unknown function (ip: 0x725fe8d00d62)
jl_apply at /cache/build/builder-amdci5-1/julialang/julia-master/src/julia.h:2228 [inlined]
do_call at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:125
eval_value at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:222
eval_stmt_value at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:173 [inlined]
eval_body at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:685
jl_interpret_toplevel_thunk at /cache/build/builder-amdci5-1/julialang/julia-master/src/interpreter.c:892
jl_toplevel_eval_flex at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1047
jl_toplevel_eval_flex at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:987
ijl_toplevel_eval at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1058
ijl_toplevel_eval_in at /cache/build/builder-amdci5-1/julialang/julia-master/src/toplevel.c:1100
eval at ./boot.jl:439 [inlined]
include_string at ./loading.jl:2741
_include at ./loading.jl:2801
include at ./Base.jl:581 [inlined]
exec_options at ./client.jl:329
_start at ./client.jl:558
jfptr__start_66271.1 at /home/blegat/.julia/juliaup/julia-nightly/lib/julia/sys.so (unknown line)
jl_apply at /cache/build/builder-amdci5-1/julialang/julia-master/src/julia.h:2228 [inlined]
true_main at /cache/build/builder-amdci5-1/julialang/julia-master/src/jlapi.c:900
jl_repl_entrypoint at /cache/build/builder-amdci5-1/julialang/julia-master/src/jlapi.c:1059
main at /cache/build/builder-amdci5-1/julialang/julia-master/cli/loader_exe.c:58
unknown function (ip: 0x725fea5e2e07)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
Allocations: 1 (Pool: 1; Big: 0); GC: 0
```

**Comment by odow** [comment]
Okay. I tidied things up. Our previous implementation was just needlessly complicated with layers of indirection.

**Pull request merged**